codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/serialize/QuerySerializerGenerator.kt

```diff
@@ -282,30 +282,38 @@ abstract class QuerySerializerGenerator(private val codegenContext: CodegenConte
         }
     }
 
-    private fun RustWriter.serializeCollection(
+    protected open fun RustWriter.serializeCollection(
         memberContext: MemberContext,
         context: Context<CollectionShape>,
+    ) {
+        serializeCollectionInner(memberContext, context, this)
+    }
+
+    protected fun serializeCollectionInner(
+        memberContext: MemberContext,
+        context: Context<CollectionShape>,
+        writer: RustWriter,
     ) {
         val flat = memberContext.shape.isFlattened()
         val memberOverride =
             when (val override = context.shape.member.getTrait<XmlNameTrait>()?.value) {
                 null -> "None"
                 else -> "Some(${override.dq()})"
             }
-        val itemName = safeName("item")
-        safeName("list").also { listName ->
-            rust("let mut $listName = ${context.writerExpression}.start_list($flat, $memberOverride);")
-            rustBlock("for $itemName in ${context.valueExpression.asRef()}") {
+        val itemName = writer.safeName("item")
+        writer.safeName("list").also { listName ->
+            writer.rust("let mut $listName = ${context.writerExpression}.start_list($flat, $memberOverride);")
+            writer.rustBlock("for $itemName in ${context.valueExpression.asRef()}") {
                 val entryName = safeName("entry")
                 Attribute.AllowUnusedMut.render(this)
-                rust("let mut $entryName = $listName.entry();")
+                writer.rust("let mut $entryName = $listName.entry();")
                 val targetShape = model.expectShape(context.shape.member.target)
-                serializeMemberValue(
+                writer.serializeMemberValue(
                     MemberContext(entryName, ValueExpression.Reference(itemName), context.shape.member),
                     targetShape,
                 )
             }
-            rust("$listName.finish();")
+            writer.rust("$listName.finish();")
         }
     }
 
```

Comment:
This looks to be doing the same thing. Why update it?

Reply:
This is working around Kotlin's limitation that an extension method cannot be overridden. `Ec2QuerySerializerGenerator.kt`, a subclass of `QuerySerializerGenerator.kt, needs to override `RustWriter.serializeCollection` so that it should only call it if an input list is NOT empty (otherwise `&ListArg=` would be rendered which ec2 query protocol doesn't like in Smithy 1.49). However, we currently couldn't do this in Kotlin if it were not for `serializeCollectionInner`:
```
override fun RustWriter.serializeCollection(
        memberContext: MemberContext,
        context: Context<CollectionShape>,
    ) {
        rustBlock("if !${context.valueExpression.asRef()}.is_empty()") {
            super.serializeCollection(memberContext, context) // cannot call super class's extension method like this
        }
```

Reply:
This is working around Kotlin's limitation that an extension method cannot be overridden. `Ec2QuerySerializerGenerator`, a subclass of `QuerySerializerGenerator`, needs to override `RustWriter.serializeCollection` so that it should only call it if an input list is NOT empty (otherwise `&ListArg=` would be rendered which ec2 query protocol doesn't like in Smithy 1.49). However, we currently couldn't do this in Kotlin if it were not for `serializeCollectionInner`:
```
override fun RustWriter.serializeCollection(
        memberContext: MemberContext,
        context: Context<CollectionShape>,
    ) {
        rustBlock("if !${context.valueExpression.asRef()}.is_empty()") {
            super.serializeCollection(memberContext, context) // cannot call super class's extension method like this
        }
```